text/0000-collection-transmute.md

```diff
@@ -0,0 +1,152 @@
+- Feature Name: collection-transmute
+- Start Date: 2019-09-05
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add a `transmute(self)` method to `Vec`, `VecDeque` and `BinaryHeap`.
+
+# Motivation
+[motivation]: #motivation
+
+The mentioned types cannot safely be transmuted by `mem::transmute`. Adding a 
+method for that purpose will hopefully discourage users to try anyway.
+
+E.g. the following code is UB:
+
+```rust
+let x = vec![0u32; 2];
+let y = unsafe { std::mem::transmute::<_, Vec<[u8; 4]>>(x) };
+```
+
+This is explained in the docs for [`Vec`], but with an unsound solution. The 
+way to do this correctly turns out surprisingly tricky:
+
+```rust
+let x = vec![0u32; 2];
+let y = unsafe {
+    let y: &mut Vec<_> = &mut *ManuallyDrop::new(x);
+    Vec::from_raw_parts(y.as_mut_ptr() as *mut [u8; 4],
+       		        y.len(),
+		        y.capacity())
+};
+```
+
+Though the code is not too large, there are a good number of things to get 
+wrong – this solution was iteratively created by soundness-knowledgeable 
+Rustaceans, with multiple wrong attempts. So this method seems like a good 
+candidate for inclusion into `std`.
+
+This also applies to `VecDeque` and `BinaryHeap`, which are implemented in 
+terms of `Vec`.
+
+[`Vec`]: https://doc.rust-lang.org/std/vec/struct.Vec.html
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The types `std::vec::Vec`, `std::collections::VecDeque` and 
+`std::collections::BinaryHeap` all get a new unsafe `transmute` method that 
+takes `self` by value and returns a new `Vec`/`VecDeque`/`BinaryHeap` with a 
+caller-chosen item type.
+
+The API might look like this (exemplified for `Vec`):
+
+```rust
+impl<T> Vec<T> {
+    /// Transmute this `Vec` to a different item type
+    ///
+    /// # Safety
+    ///
+    /// Calling this function requires the target item type be compatible with
+    /// `Self::Item` (see [`mem::transmute`]).
```

Comment:
I think it would be better to just say something like:

> Safe if `transmute::<T,I>` is safe.

instead of using words like "compatible" which might not have a clear meaning for everybody.

Reply:
cc @rust-lang/wg-unsafe-code-guidelines